Add ability to partially match Object and Arrays with the new match keyword (#60)
kuzzleio/koncorde: doc/terms/index.md

}
```

## match

Test if properties of the filter is matching with the properties of the document.

> If A and B both are Scalars, checks that they're equal.

Reviewer: Can you use the term `filter` for A and `subject` for B? (it's a naming proposal for more clarity and I'm open to other suggestions)

If A and B both are Objects, checks that every property of A is matching a property of B.

> If A and B both are Arrays, checks that every values of A has a value in B matching.

Reviewer: ```suggestion
If A and B both are Arrays, checks that every values of A is also matching a value in B.
```

> This process is recursive, meaning that Sub Objects and Arrays are expected to match each other and not be strictly equal.

Reviewer: ```suggestion
This process is recursive, meaning that nested objects and arrays are expected to match each other on the same principle.
```

:::warning
Matching Arrays is O(n * m) with n being the number of elements in A and m the numbers of elements in B

Matching Objects is O(min(n,m)) with n being the number of properties in A and m the numbers of properties in B

Matching Scalars is O(1)

> Keep in mind that since this process is recursive on Sub Objects and Arrays the complexity can increase.

Reviewer: ```suggestion
Keep in mind that since this process is recursive on nested objects and arrays, the complexity can increase.
```

:::

### Syntax

```
match: {
  <field name>: <value>,
  <field name>: <value>
  ...
}
```

### Example

#### Array Matching
Given the following filter:

```js
{
  "match": {
    "foo": [1, 2, 3]
  }
}
```

It will match this document, because all values from the filter `foo` property are contained in the document `foo` property.

```js
{
  "foo": [0, 1, 4, 2, 3]
}
```

> But will not match this document, because the document `foo` property is missing some values.

Reviewer: ```suggestion
But will not match this document, because the `foo` array is missing the value `3` from the filter.
```

```js
{
  "foo": [0, 0, 1, 2]
}
```

#### Object Matching

Given the following filter:

```js
{
  "match": {

>     "foo": {
>       "a": 5
>     }

Reviewer: ```suggestion
    "foo": { "a": 5 }
```

  }
}

```

> It will match this document, because all properties from the filter `foo` Object can be matched against the document `foo` property.

Reviewer: ```suggestion
It will match this document, because all properties from the `foo` object of the filter can be matched with corresponding properties in the `foo` object of the document.
```

```js
{
  "foo": {
    "a": 5,
    "b": 8,
    "c": "something"
  }
}
```

> But will not match this document, because the document `foo` Object is missing some properties or they have not the same values as the properties from the `foo` Object of the filter.

Reviewer: ```suggestion
But will not match this document, because the `foo` object of the document is missing some properties or they have not the same values as the properties from the `foo` object of the filter.
```
